### PR TITLE
Align with draft-irtf-cfrg-vdaf-12 (*)

### LINKED ISSUE #34: Harmonize VIDPF with changes to IDPF in draft-irtf-cfrg-vdaf-12
We would like these to be aligned as possible. We're still optimizing IDPF here and there, e.g., https://github.com/cfrg/draft-irtf-cfrg-vdaf/pull/303. We should decide if we're OK with these optimizations.

### PR DESCRIPTION
**Pull request by cjpatton**
Closes #34.

* Plumb the application context string into the domain separation tag of each XOF invocation.
* Change the VIDPF indices (inputs and prefixes) to `tuple[bool, ...]`, i.e., a sequence of bools.

### CODE REVIEW

File: draft-mouris-cfrg-mastic.md [hunk 1]
```diff
@@ -521,14 +513,15 @@ def eval(self,
...
                         nonce,
                         child_idx,
                     )
                     prefix_tree_share[child_idx] = child
 
     # Compute the aggregator's share of `beta`.
-    w0 = prefix_tree_share[PrefixTreeIndex(0, 0)].w
-    w1 = prefix_tree_share[PrefixTreeIndex(1, 0)].w
+    w0 = prefix_tree_share[PrefixTreeIndex((False,))].w
+    w1 = prefix_tree_share[PrefixTreeIndex((True,))].w
```
Comment (jimouris) [inline draft-mouris-cfrg-mastic.md:524]: Shouldn't this be:
```suggestion
    w0 = prefix_tree_share[PrefixTreeIndex((False, False,))].w
    w1 = prefix_tree_share[PrefixTreeIndex((True, False,))].w
```

File: draft-mouris-cfrg-mastic.md [hunk 2]
```diff
@@ -347,8 +341,9 @@ each level of the tree, and the secret keys, one for each Aggregator.
 
 ~~~ python
 def gen(self,
-        alpha: int,
+        alpha: tuple[bool, ...],
         beta: list[F],
+        ctx: bytes,
```
Comment (jimouris) [inline draft-mouris-cfrg-mastic.md:346]: Add a comment about what `ctx` is? I couldn't figure it out from the changes

File: poc/mastic.py [hunk 3]
```diff
@@ -283,13 +291,14 @@ def prep_shares_to_prep(
...
         # Confirm the FLP joint randomness was computed properly.
-        prep_msg = self.joint_rand_seed([
+        prep_msg = self.joint_rand_seed(ctx, [
             joint_rand_part_0,
             joint_rand_part_1,
         ])
         return prep_msg
 
     def prep_next(self,
+                  _ctx: bytes,
```
Comment (jimouris) [inline poc/mastic.py:301]: why pass it here?

File: poc/tests/test_mastic.py [hunk 4]
```diff
@@ -10,6 +10,7 @@
 
 class TestValidAggParams(unittest.TestCase):
     def test(self):
+        # XXX
```
Comment (jimouris) [inline poc/tests/test_mastic.py:13]: ?

File: poc/vidpf.py [hunk 5]
```diff
@@ -94,8 +103,9 @@ def __init__(self, field: type[F], bits: int, value_len: int):
         self.VALUE_LEN = value_len
 
     def gen(self,
-            alpha: int,
+            alpha: tuple[bool, ...],
```
Comment (jimouris) [inline poc/vidpf.py:106]: Why the dots? Isn't `alpha: tuple[bool]` sufficient?

File: poc/vidpf.py [hunk 6]
```diff
@@ -254,14 +261,15 @@ def eval(self,
                             node,
                             onehot_proof,
                             correction_words[i],
+                            ctx,
```
Comment (jimouris) [inline poc/vidpf.py:264]: At this point I have understood what `ctx` :) but the name `ctx` is not very descriptive IMO. I understand that we have it as is to match the VDAF draft but could we have some comments in various places we use it as the name doesn't really imply what it is? (does it stand for `context`?)

File: poc/vidpf.py [hunk 7]
```diff
@@ -254,14 +261,15 @@ def eval(self,
...
                             nonce,
                             child_idx,
                         )
                         prefix_tree_share[child_idx] = child
 
         # Compute the aggregator's share of `beta`.
-        w0 = prefix_tree_share[PrefixTreeIndex(0, 0)].w
-        w1 = prefix_tree_share[PrefixTreeIndex(1, 0)].w
+        w0 = prefix_tree_share[PrefixTreeIndex((False,))].w
```
Comment (jimouris) [inline poc/vidpf.py:271]:
```suggestion
        w0 = prefix_tree_share[PrefixTreeIndex((False, False,))].w
```
?

File: poc/vidpf.py [hunk 8]
```diff
@@ -332,8 +342,8 @@ def eval_next(self,
         # Implementation note: the conditional addition should be
         # replaced with a constant-time select in practice in order to
         # reduce leakage via timing side channels.
-        (next_seed, w) = self.convert(s[keep], nonce)  # [MST24]: s^i,W^i
-        next_ctrl = t[keep]  # [MST24]: t'^i
+        (next_seed, w) = self.convert(s[keep], ctx, nonce)
+        next_ctrl = t[keep]  # [MST24]: s^i,W^i, t'^i
```
Comment (jimouris) [inline poc/vidpf.py:346]:
```suggestion
        next_ctrl = t[keep]  # [MST24]: s^i, W^i, t'^i
```

**Review by jimouris — APPROVED** [review]

File: poc/mastic.py [hunk 9]
```diff
@@ -19,9 +19,9 @@
 F = TypeVar("F", bound=NttField)
 
 MasticAggParam: TypeAlias = tuple[
-    int,            # level
-    Sequence[int],  # prefixes
-    bool,           # whether to do the weight check
+    int,                      # level
+    tuple[tuple[bool, ...]],  # prefixes
```
Comment (TrustworthyComputing) [inline poc/mastic.py:23]: why are we using a nested tuple of booleans instead of an int sequence?

File: poc/vidpf.py (hunk 6 above)
Comment (TrustworthyComputing) [reply poc/vidpf.py:264]: Should we enforce any constraints to the `ctx` size? The VDAF draft mentions the clients and aggregators need to agree on the string. Would it make sense to require that this is a hash of the context instead?

File: poc/vidpf.py [hunk 10]
```diff
@@ -254,14 +261,15 @@ def eval(self,
...
                             nonce,
                             child_idx,
                         )
                         prefix_tree_share[child_idx] = child
 
         # Compute the aggregator's share of `beta`.
-        w0 = prefix_tree_share[PrefixTreeIndex(0, 0)].w
-        w1 = prefix_tree_share[PrefixTreeIndex(1, 0)].w
+        w0 = prefix_tree_share[PrefixTreeIndex((False,))].w
+        w1 = prefix_tree_share[PrefixTreeIndex((True,))].w
```
Comment (TrustworthyComputing) [inline poc/vidpf.py:272]: same as L271

File: draft-mouris-cfrg-mastic.md [hunk 11]
```diff
@@ -825,36 +833,38 @@ The complete algorithm is listed below:
...
+        measurement: tuple[tuple[bool, ...], W],
         nonce: bytes,
         rand: bytes,
 ) -> tuple[MasticPublicShare, list[MasticInputShare]]:
     (vidpf_rand, rand) = front(self.vidpf.RAND_SIZE, rand)
     (prove_rand_seed, rand) = front(self.xof.SEED_SIZE, rand)
     (leader_seed, rand) = front(self.xof.SEED_SIZE, rand)
     (helper_seed, rand) = front(self.xof.SEED_SIZE, rand)
+    assert len(rand) == 0  # REMOVE ME
```
Comment (TrustworthyComputing) [inline draft-mouris-cfrg-mastic.md:845]: ?

**Review by jimouris — APPROVED** [review]

**Pull request merged**